README.md
<!-- deno-fmt-ignore-file -->

JSON Hash
=========

[![Latest version][Tag badge]][Deno module]
[![LGPL 3.0][License badge]](./LICENSE)
[![Deno Doc (API references)][Deno Doc badge]][Deno Doc]
[![GitHub Actions][GitHub Actions status badge]][GitHub Actions]
[![Codecov][Codecov badge]][Codecov]

This [Deno][][^1] package contains several utilities to deal JSON data with
cryptography.  See the below sections for details.

[^1]: It is open to expand its target runtimes including [web browsers in
      particular][1].

[Tag badge]: https://img.shields.io/github/v/tag/dahlia/json-hash
[Deno module]: https://deno.land/x/json_hash
[License badge]: https://img.shields.io/github/license/dahlia/json-hash
[Deno Doc]: https://doc.deno.land/https://deno.land/x/json_hash/mod.ts
[Deno Doc badge]: https://img.shields.io/badge/api-deno%20doc-blue
[GitHub Actions]: https://github.com/dahlia/json-hash/actions/workflows/test.yaml
[GitHub Actions status badge]: https://github.com/dahlia/json-hash/actions/workflows/test.yaml/badge.svg
[Codecov badge]: https://codecov.io/gh/dahlia/json-hash/branch/main/graph/badge.svg?token=rCNDUvONJv
[Codecov]: https://codecov.io/gh/dahlia/json-hash
[Deno]: https://deno.land/
[1]: https://github.com/dahlia/json-hash/issues/2


[*canon.ts*][canon.ts]: JSON normalizer compliant with JCS
----------------------------------------------------------

Cryptographic operations like hashing and signing need the data to be
expressed in an invariant format so that the operations are reliably
repeatable.  Even if two messages are similar, only a single byte of difference
causes totally different digests or signatures.

However, JSON allows stylistic freedom on their representations.
For example, the following representations all encodes the same JSON entity:

~~~ jsonl
{ "foo": "bar", "baz": [1, 2, 3] }
{ "baz": [1, 2, 3], "foo": "bar" }
{"foo":"bar","baz":[1,2,3]}
{ "\u0066\u006f\u006f": "\u0062\u0061\u0072", "\u0062\u0061\u007a": [1, 2, 3] }
~~~~

In order to hash or sign JSON data, they should be normalized first.
That's why JSON Canonicalization Scheme (JCS) was proposed in [RFC 8785].
JCS allows only a single representation for each possible JSON entity.
For example, the JSON entity the above multiple representations encode can
be represented into only the below single form:

~~~ json
{"baz":[1,2,3],"foo":"bar"}
~~~

The [*canon.ts*][canon.ts] module implements JCS, which completely complies
with [RFC 8785].  Here's some example:

~~~ typescript
import { canonicalize } from "https://deno.land/x/json_hash/canon.ts";
import { assertEquals } from "https://deno.land/std/testing/asserts.ts";

assertEquals(
  canonicalize({ foo: "bar", baz: [1, 2, 3] }),
  canonicalize({
    "\x62\x61\x7a": [1, 2, 3],
    "\x66\x6f\x6f": "\x62\x61\x72",
  }),
);
~~~

[canon.ts]: https://doc.deno.land/https://deno.land/x/json_hash/canon.ts
[RFC 8785]: https://tools.ietf.org/html/rfc8785


[*digest.ts*][digest.ts]: Hashing JSON entities
-----------------------------------------------

This module is a thin wrapper around the above `canonicalize()` function and
the curated collection of cryptographic hash algorithms:

~~~~ typescript
import { digest } from "https://deno.land/x/json_hash/digest.ts";
import { toHex } from "https://deno.land/x/json_hash/hex.ts";

const data = { foo: "bar", baz: [ 1, 2, 3 ]};
const hash: Uint8Array = await digest("BLAKE3", data);
console.log(toHex(hash));
// dec9c1a89be824103c812b7ace381263335cd6f421a4d0f4dd407a4d3335189c
~~~~

The `digest()` function guarantees that it will return the equivalent hash
digest (in `Uint8Array`) for the same hash algorithm and the equivalent JSON
entity.  It means you don't have to care about the order of how object keys
occurs or how characters in string are encoded.

Also note that the above example uses [BLAKE3], which is unsupported by
Web Crypto API (as of January 2022).  Powered by Deno's [std/crypto] module,
which is baked into Web Assembly, it provides much wider range of hash
algorithms than Web Crypto API.

[digest.ts]: https://doc.deno.land/https://deno.land/x/json_hash/digest.ts
[BLAKE3]: https://github.com/BLAKE3-team/BLAKE3
[std/crypto]: https://deno.land/std@0.120.0/crypto#supported-algorithms


[*merkle.ts*][merkle.ts]: Dealing JSON with Merkle tree
-------------------------------------------------------

If you need to track changes of large JSON trees this module could help you.
It provides `merkle()` function and `MerkleHash<T>` class which enables you to
build [Merkle tree]s.

Unlike `digest()` function from *digest.ts* module, `merkle()` function doesn't
digest the entire JSON data at once, but digests each small entity of the tree
and digests them recursively.  For example:

~~~ typescript
import { MerkleHash, merkle } from "https://deno.land/x/json_hash/merkle.ts";
import { assert } from "https://deno.land/std/testing/asserts.ts";

const a: MerkleHash<"BLAKE3"> = await merkle("BLAKE3", [1, 2, 3]);
const b: MerkleHash<"BLAKE3"> = await merkle("BLAKE3", [
  await merkle("BLAKE3", 1),
  await merkle("BLAKE3", 2),
  await merkle("BLAKE3", 3),
]);
assert(a.equals(b));

const c: MerkleHash<"BLAKE3"> = await merkle(
  "BLAKE3",
  { foo: "bar", baz: [1, 2, 3] }
);
const d: MerkleHash<"BLAKE3"> = await merkle(
  "BLAKE3",
  { foo: await merkle("BLAKE3", "bar"), baz: a }
);
assert(c.equals(d));

const e = await merkle(c);
assert(e.equals(c));
~~~

Note that `merkle()` returns `MerkleHash<T>` (instead of `Uint8Array`),
and also takes a JSON tree mixed with `MerkleHash<T>` values.  In this module,
such mixed trees are called `MerkleTree<T>`.[^2]  The `merkle()` function
behaves like below:

 -  `merkle()` derives `MerkleHash<T>` from `MerkleTree<T>`.
 -  `merkle()` makes no distinction between a `MerkleTree<T>` and
    `MerkleHash<T>` derived from the same tree.

In other words, the idea this module implements is very simple: hierarchical
hash verification of JSON trees.  If you still don't get it well I recommend
you to read *[Diving into Merkle Trees]* by Pedro Tavares.

[^2]: It is a compiled-time abstract type, and the type parameter `T` represents
      the hash algorithm, which corresponds to `MerkleHash<T>`'s type parameter.

[merkle.ts]: https://doc.deno.land/https://deno.land/x/json_hash/digest.ts
[Merkle tree]: https://en.wikipedia.org/wiki/Merkle_tree
[Diving into Merkle Trees]: https://ordep.dev/posts/diving-into-merkle-trees


[*mod.ts*][Deno doc]: Façade
----------------------------

This module re-exports everything in the above files.

See also [Deno Doc] for the complete API references.


Changelog
---------

See [CHANGES.md](CHANGES.md) file.


License
-------

Distributed under [LGPL 3.0] or later.

[LGPL 3.0]: https://www.gnu.org/licenses/lgpl-3.0.html


CHANGES.md
<!-- deno-fmt-ignore-file -->

JSON Hash Changelog
===================

Version 0.2.0
-------------

Released on July 29, 2022.

 -  Added `MerkleHash#compareTo()` method.  [[#4]]
 -  `new MerkleHash()` constructor became to throw `TypeError` instead of
    `Error` if its second argument is a wrong type.
 -  `MerkleHash#equals()` method became to return `false` instead of throwing
    `TypeError` if its operand is not a `MerkleHash` instance.

[#4]: https://github.com/dahlia/json-hash/issues/4


Version 0.1.0
-------------

Released on January 11, 2022.


hex_test.ts
import {
  assertEquals,
  assertThrows,
} from "https://deno.land/std@0.120.0/testing/asserts.ts";
import { fromHex, toHex } from "./hex.ts";

Deno.test("toHex()", () => {
  assertEquals(toHex(Uint8Array.of()), "");
  assertEquals(toHex(Uint8Array.of(0x12, 0x34, 0x56, 0x78)), "12345678");
  assertEquals(toHex(Uint8Array.of(0xfe, 0xdc, 0xba, 0x00)), "fedcba00");
});

Deno.test("fromHex()", () => {
  assertEquals(fromHex(""), Uint8Array.of());
  assertEquals(fromHex("12345678"), Uint8Array.of(0x12, 0x34, 0x56, 0x78));
  assertEquals(fromHex("fedcba00"), Uint8Array.of(0xfe, 0xdc, 0xba, 0x00));
  assertEquals(fromHex("ABCDEF00"), Uint8Array.of(0xab, 0xcd, 0xef, 0x00));
  assertThrows(() => fromHex("a"), RangeError, "length");
  assertThrows(() => fromHex("abc"), RangeError, "length");
  assertThrows(() => fromHex("zz"), RangeError);
});


canon_test.ts
import { canonicalize } from "./canon.ts";
import { fromHex } from "./hex.ts";
import {
  assertEquals,
  AssertionError,
} from "https://deno.land/std@0.120.0/testing/asserts.ts";

Deno.test("canonicalize(null)", () => {
  assertEquals(canonicalize(null), "null");
});

Deno.test("canonicalize(boolean)", () => {
  assertEquals(canonicalize(true), "true");
  assertEquals(canonicalize(false), "false");
});

function doubleBE(hex: string): number {
  const buffer = fromHex(hex);
  const view = new DataView(buffer.buffer);
  return view.getFloat64(0, false);
}

// https://tools.ietf.org/id/draft-rundgren-json-canonicalization-scheme-00.html#rfc.section.3.2.2.3  // editorconfig-checker-disable-line
Deno.test("canonicalize(number)", () => {
  assertEquals(
    canonicalize(doubleBE("4340000000000001")),
    "9007199254740994",
  );
  assertEquals(
    canonicalize(doubleBE("4340000000000002")),
    "9007199254740996",
  );
  assertEquals(canonicalize(doubleBE("444b1ae4d6e2ef50")), "1e+21");
  assertEquals(canonicalize(doubleBE("3eb0c6f7a0b5ed8d")), "0.000001");
  assertEquals(
    canonicalize(doubleBE("3eb0c6f7a0b5ed8c")),
    "9.999999999999997e-7",
  );
  assertEquals(canonicalize(doubleBE("8000000000000000")), "0");
  assertEquals(canonicalize(doubleBE("7fffffffffffffff")), "null");
  assertEquals(canonicalize(doubleBE("7ff0000000000000")), "null");
  assertEquals(canonicalize(doubleBE("fff0000000000000")), "null");
});

Deno.test("canonicalize(string)", () => {
  assertEquals(canonicalize(""), `""`);
  assertEquals(
    canonicalize("predefined control chars: \b\t\n\f\r"),
    `"predefined control chars: \\b\\t\\n\\f\\r"`,
  );
  assertEquals(
    canonicalize("control chars: \u0000\u0001\u0002\u0003\u001e\u001f"),
    `"control chars: \\u0000\\u0001\\u0002\\u0003\\u001e\\u001f"`,
  );
  assertEquals(
    canonicalize(`quotes and backslashes: "\\`),
    `"quotes and backslashes: \\"\\\\"`,
  );
  assertEquals(
    canonicalize("other chars: ABC xyz 가나다 123 甲乙丙"),
    `"other chars: ABC xyz 가나다 123 甲乙丙"`,
  );
});

Deno.test("canonicalize(unknown[])", () => {
  assertEquals(canonicalize([]), "[]");
  assertEquals(canonicalize([1, 2, 3]), "[1,2,3]");
  assertEquals(canonicalize([1, 2, 3, null]), "[1,2,3,null]");
  assertEquals(canonicalize(["asdf", true, false]), `["asdf",true,false]`);
});

Deno.test("canonicalize(object)", () => {
  assertEquals(canonicalize({}), "{}");
  assertEquals(canonicalize({ a: 1, b: 2, c: 3 }), `{"a":1,"b":2,"c":3}`);
  assertEquals(
    canonicalize({
      가나다: null,
      "\r\n": true,
      "": {},
      '"\\': [{ a: 1, b: 2 }],
    }),
    `{"":{},"\\r\\n":true,"\\"\\\\":[{"a":1,"b":2}],"가나다":null}`,
  );
});

Deno.test("canonicalize() [JCS compliance]", async () => {
  const jcsTestData = "jcs_spec/testdata";
  try {
    for await (const f of Deno.readDir(`${jcsTestData}/input`)) {
      if (!f.isFile) continue;
      const input = await Deno.readTextFile(`${jcsTestData}/input/${f.name}`);
      const inputTree = JSON.parse(input);
      const output = canonicalize(inputTree);
      const expected = await Deno.readTextFile(
        `${jcsTestData}/output/${f.name}`,
      );
      if (output !== expected) {
        console.error(f.name);
      }
      assertEquals(output, expected);
    }
  } catch (e) {
    if (e instanceof Deno.errors.NotFound) {
      throw new AssertionError(
        "No JCS test suite found.  Please check if submodules are initialized.",
      );
    }

    throw e;
  }
});


digest_test.ts
import { digest } from "./digest.ts";
import { toHex } from "./hex.ts";
import { assertEquals } from "https://deno.land/std@0.120.0/testing/asserts.ts";

Deno.test("digest()", async () => {
  const a1 = { foo: 1, bar: { x: 1, y: 2 } };
  const a2 = { bar: { y: 2, x: 1 }, foo: 1 };
  const b = { bar: { x: 1, y: 1 }, foo: 1 };
  const hashA1 = await digest("SHA-256", a1);
  assertEquals(
    toHex(hashA1),
    "fc9fcf9bc651901c1b870fa315676818fd95407a0444f73b9f931858978addf4",
  );
  const hashA2 = await digest("SHA-256", a2);
  assertEquals(hashA2, hashA1);
  const hashB = await digest("SHA-256", b);
  assertEquals(
    toHex(hashB),
    "9e92e99663a38271f9476f337a835ad88294c6afe9416e08db0e7285e5b93a07",
  );
});


hex.ts
const hexTable: string[] = [];
for (let i = 0; i < 0x100; ++i) {
  hexTable.push(i.toString(16).padStart(2, "0"));
}

/**
 * Turns a {@link Uint8Array} into a hexadecimal string.
 * @param bytes The bytes to convert.
 * @returns The hexadecimal string.  The string will be lowercase.
 */
export function toHex(bytes: Uint8Array): string {
  const result = [];
  for (let i = 0; i < bytes.length; ++i) {
    result.push(hexTable[bytes[i]]);
  }
  return result.join("");
}

/**
 * Turns a hexadecimal string into a {@link Uint8Array}.
 * @param hex The hexadecimal string to convert.  It is case-insensitive.
 * @returns The bytes.
 * @throws {RangeError} Thrown if the string is not a valid hexadecimal string.
 */
export function fromHex(hex: string): Uint8Array {
  if (hex.length % 2 !== 0) {
    throw new RangeError("Invalid hex string; length must be even.");
  } else if (!hex.match(/^[0-9a-fA-F]*$/)) {
    throw new RangeError("Invalid hex string; must be hexadecimal.");
  }
  const result = new Uint8Array(hex.length / 2);
  for (let i = 0; i < hex.length; i += 2) {
    result[i / 2] = parseInt(hex.substring(i, i + 2), 16);
  }
  return result;
}


canon.ts
/**
 * This module provides canonicalization of JSON trees.  This complies with
 * RFC 8785, also known as JCS (JSON Canonicalization Scheme).
 * @license LGPL-3.0-or-later
 */

/**
 * Represents a JSON tree.
 */
export type Tree =
  | null
  | boolean
  | number
  | string
  | TreeArray
  | TreeObject;

// deno-lint-ignore no-empty-interface
interface TreeArray extends Array<Tree> {
}

interface TreeObject {
  [_: string]: Tree;
}

function serialize(tree: Tree, buffer: string[]): void {
  if (tree == null || typeof tree !== "object") {
    // ES6 JSON partially complies with RFC 8785 (JCS: JSON Canonicalization
    // Scheme) except for objects and arrays:
    buffer.push(JSON.stringify(tree));
  } else if (Array.isArray(tree)) {
    buffer.push("[");
    let notFirst = false;
    for (const item of tree) {
      if (notFirst) {
        buffer.push(",");
      }
      notFirst = true;
      serialize(item, buffer);
    }
    buffer.push("]");
  } else {
    // Object keys must be sorted in JSON.
    const keys = Object.keys(tree);
    keys.sort();
    buffer.push("{");
    let notFirst = false;
    for (const key of keys) {
      if (notFirst) {
        buffer.push(",");
      }
      notFirst = true;
      buffer.push(JSON.stringify(key));
      buffer.push(":");
      serialize(tree[key], buffer);
    }
    buffer.push("}");
  }
}

/**
 * Serializes a tree into a JSON string which complies with RFC 8785
 * (JCS: JSON Canonicalization Scheme).
 * @param tree The tree to serialize.
 * @returns The canonicalized JSON string.
 */
export function canonicalize(tree: Tree): string {
  const buffer: string[] = [];
  serialize(tree, buffer);
  return buffer.join("");
}


digest.ts
/**
 * This module provides cryptographic hash digests of JSON trees.
 * It guarantee that `digest()` function always returns the same digest
 * for the equivalent JSON trees.  It means you don't have to care about
 * the order of how object keys occurs or how characters in string are encoded.
 * It also provides various hash algorithms; see also the docs of Deno's
 * std/crypto module.
 * @license LGPL-3.0-or-later
 */
import { crypto, type DigestAlgorithmType } from "./crypto.ts";
import { canonicalize, type Tree } from "./canon.ts";

/**
 * Digests a tree using the specified hash algorithm.
 * @param algorithm The hash algorithm to use.
 * @param data The JSON tree to digest.
 * @returns The hash digest.
 * @throws {DOMException} Thrown when the algorithm is not supported.
 */
export async function digest(
  algorithm: DigestAlgorithmType,
  data: Tree,
): Promise<Uint8Array> {
  const canon = canonicalize(data);
  const canonBytes = new TextEncoder().encode(canon);
  const digestBuffer = await crypto.subtle.digest(algorithm, canonBytes);
  return new Uint8Array(digestBuffer);
}

export { type DigestAlgorithmType, type Tree };


merkle.ts
/**
 * This module provides hash digests of JSON Merkle trees.  It can be used for
 * efficient diff of two large trees or updating a deep leaf in a large tree.
 * @license LGPL-3.0-or-later
 */
import * as ascii85 from "https://deno.land/std@0.120.0/encoding/ascii85.ts";
import { Tree } from "./canon.ts";
import { crypto, type DigestAlgorithmType } from "./crypto.ts";
import { fromHex, toHex } from "./hex.ts";

/**
 * Represents a JSON Merkle tree.  It is equivalent to {@link Tree}, except that
 * it can contain {@link MerkleHash}es instead of actual JSON values.
 */
export type MerkleTree<T extends DigestAlgorithmType> =
  | Tree
  | MerkleTreeArray<T>
  | MerkleTreeObject<T>
  | MerkleHash<T>;

// deno-lint-ignore no-empty-interface
interface MerkleTreeArray<T extends DigestAlgorithmType>
  extends Array<MerkleTree<T>> {
}

interface MerkleTreeObject<T extends DigestAlgorithmType> {
  [_: string]: MerkleTree<T>;
}

/**
 * Represents a hash of a Merkle tree.
 */
export class MerkleHash<T extends DigestAlgorithmType> {
  static readonly #MoveFromInternalBuffer: unknown = {};
  readonly #hashBuffer: Uint8Array;

  /**
   * The hash algorithm used to create this digest.
   */
  readonly algorithm: T & DigestAlgorithmType;

  /**
   * Creates a new {@link MerkleHash} from a {@link Uint8Array}.
   * @param algorithm The used hash algorithm.
   * @param hashBuffer A buffer of digest bytes.
   */
  constructor(
    algorithm: T & DigestAlgorithmType,
    hashBuffer: ArrayBuffer | Uint8Array | number[],
    private _moveFromInternalBuffer?: unknown,
  ) {
    if (hashBuffer instanceof ArrayBuffer) {
      this.#hashBuffer =
        _moveFromInternalBuffer === MerkleHash.#MoveFromInternalBuffer
          ? new Uint8Array(hashBuffer)
          : new Uint8Array(new Uint8Array(hashBuffer));
    } else if (hashBuffer instanceof Uint8Array || Array.isArray(hashBuffer)) {
      this.#hashBuffer = new Uint8Array(hashBuffer);
    } else {
      throw new TypeError(`Unexpected type: ${typeof hashBuffer}.`);
    }
    this.algorithm = algorithm;
  }

  /**
   * Derives a {@link MerkleHash} from input data.
   * @param algorithm The hash algorithm to use.
   * @param data Input bytes to digest.
   * @returns The derived {@link MerkleHash}.
   */
  static async deriveFrom<T extends DigestAlgorithmType>(
    algorithm: T & DigestAlgorithmType,
    data: BufferSource | AsyncIterable<BufferSource> | Iterable<BufferSource>,
  ): Promise<MerkleHash<T>> {
    const digest = await crypto.subtle.digest(algorithm, data);
    return new MerkleHash(
      algorithm,
      digest,
      MerkleHash.#MoveFromInternalBuffer,
    );
  }

  /**
   * Converts a hexadecimal string into a {@link MerkleHash}.
   * @param algorithm The used hash algorithm.
   * @param hex The hexadecimal representation of a hash.
   * @returns A {@link MerkleHash} object.
   */
  static fromHex<T extends DigestAlgorithmType>(
    algorithm: T & DigestAlgorithmType,
    hex: string,
  ): MerkleHash<T> {
    const bytes = fromHex(hex);
    return new MerkleHash(
      algorithm,
      bytes.buffer,
      MerkleHash.#MoveFromInternalBuffer,
    );
  }

  /**
   * Shows the hexadecimal representation of the hash.
   */
  get hex(): string {
    return toHex(this.#hashBuffer);
  }

  /**
   * Shows the base85 (RFC 1924) representation of the hash.
   */
  get base85(): string {
    return ascii85.encode(this.#hashBuffer, {
      "standard": "RFC 1924",
      delimiter: false,
    });
  }

  /**
   * Checks if two {@link MerkleHash} objects are equal.
   * @param other The other {@link MerkleHash} to compare.
   * @returns `true` if the hashes are equal, `false` otherwise.
   */
  equals(other: MerkleHash<T>): boolean {
    if (!(other instanceof MerkleHash) || this.algorithm !== other.algorithm) {
      return false;
    }
    const thisBuffer = this.#hashBuffer;
    const otherBuffer = other.#hashBuffer;
    const length = thisBuffer.length;
    if (length !== otherBuffer.length) return false;
    for (let i = 0; i < length; i++) {
      if (thisBuffer.at(i) !== otherBuffer.at(i)) return false;
    }
    return true;
  }

  /**
   * Compares two {@link MerkleHash} objects.  The comparison is done by
   * comparing the hashes in byte order, that is, it's lexicographic.
   * @param other The other {@link MerkleHash} to compare.
   * @returns `-1` if this hash goes before the other hash, `0` if they are
   *          equal, and `1` if this hash goes after the other hash.
   * @throws {TypeError} Thrown if the other hash is a different kind of
   *                     {@link MerkleHash}; for example, different hash
   *                     algorithms.
   */
  compareTo(other: MerkleHash<T>): number {
    if (!(other instanceof MerkleHash)) {
      throw new TypeError("Expected a merkle hash.");
    }
    const thisBuffer = this.#hashBuffer;
    const otherBuffer = other.#hashBuffer;
    const length = thisBuffer.length;
    if (length !== otherBuffer.length) {
      throw new TypeError(
        `Cannot compare merkle hashes with different sizes.`,
      );
    } else if (this.algorithm !== other.algorithm) {
      throw new TypeError(
        `Cannot compare merkle hashes with different algorithms.`,
      );
    }
    for (let i = 0; i < length; i++) {
      const a = thisBuffer.at(i);
      const b = otherBuffer.at(i);
      if (a !== b) return a! - b!;
    }
    return 0;
  }

  /**
   * Returns the {@link Uint8Array} representation of the hash.
   * @returns A copy of the {@link Uint8Array} representation of the hash.
   */
  toUint8Array(): Uint8Array {
    return new Uint8Array(this.#hashBuffer);
  }

  /**
   * Shows the algorithm and hexadecimal representation of the hash.
   * @returns The algorithm and hexadecimal representation of the hash.
   */
  toString(): string {
    return `${this.algorithm.toString()} ${this.hex}`;
  }

  [Symbol.for("Deno.customInspect")](): string {
    return `MerkleHash ${this.toString()}`;
  }
}

const utf8Encoder = new TextEncoder();

// editorconfig-checker-disable
/**
 * Derives a {@link MerkleHash} from input data.
 *
 * It treats {@link MerkleHash}es in the input tree as like JSON values that
 * those hashes represent.  For example, the following code prints 5 rows of
 * the exactly same hashes:
 *
 * ```ts
 * import { MerkleTree, merkle } from "./merkle.ts";
 * const tree = [
 *   {
 *     title: "The Catcher in the Rye",
 *     author: "J.D. Salinger",
 *     content: "Very long content...",
 *   },
 *   {
 *    title: "The Great Gatsby",
 *    author: "F. Scott Fitzgerald",
 *    content: "Very long content...",
 *   },
 * ];
 *
 * const tree2: MerkleTree<"SHA3-256"> = [
 *   {
 *     ...tree[0],
 *     content: await merkle("SHA3-256", tree[0].content),
 *   },
 *   tree[1],
 * ];
 *
 * const tree3: MerkleTree<"SHA3-256"> = [
 *   await merkle("SHA3-256", tree[0]),
 *   tree[1],
 * ];
 *
 * const tree4: MerkleTree<"SHA3-256"> = [
 *   await merkle("SHA3-256", tree[0]),
 *   await merkle("SHA3-256", tree[1]),
 * ];
 *
 * const merkleRoot = await merkle("SHA3-256", tree);
 *
 * console.log(await merkle("SHA3-256", tree));
 * console.log(await merkle("SHA3-256", tree2));
 * console.log(await merkle("SHA3-256", tree3));
 * console.log(await merkle("SHA3-256", tree4));
 * console.log(await merkle("SHA3-256", merkleRoot));
 *
 * // prints:
 * // MerkleHash SHA3-256 ede1427e0292877c45436591ac28139e0a8a1c23cf5d2036ffad6067bbc8c15c
 * // MerkleHash SHA3-256 ede1427e0292877c45436591ac28139e0a8a1c23cf5d2036ffad6067bbc8c15c
 * // MerkleHash SHA3-256 ede1427e0292877c45436591ac28139e0a8a1c23cf5d2036ffad6067bbc8c15c
 * // MerkleHash SHA3-256 ede1427e0292877c45436591ac28139e0a8a1c23cf5d2036ffad6067bbc8c15c
 * // MerkleHash SHA3-256 ede1427e0292877c45436591ac28139e0a8a1c23cf5d2036ffad6067bbc8c15c
 * ```
 *
 * @param algorithm The hash algorithm to use.
 * @param tree The input JSON data.
 * @returns The merkle root hash of the given JSON tree.
 */
// editorconfig-checker-enable
export async function merkle<T extends DigestAlgorithmType>(
  algorithm: T & DigestAlgorithmType,
  tree: MerkleTree<T>,
): Promise<MerkleHash<T>> {
  if (tree instanceof MerkleHash) return tree;
  let canon: string;
  if (tree == null || typeof tree !== "object") {
    // ES6 JSON partially complies with RFC 8785 (JCS: JSON Canonicalization
    // Scheme) except for objects and arrays:
    canon = JSON.stringify(tree);
  } else if (Array.isArray(tree)) {
    const buffer: string[] = [];
    buffer.push("[");
    let notFirst = false;
    for (const item of tree) {
      buffer.push(notFirst ? ",'" : "'");
      notFirst = true;
      const itemHash = await merkle(algorithm, item);
      buffer.push(itemHash.base85);
      buffer.push("'");
    }
    buffer.push("]");
    canon = buffer.join("");
  } else {
    // Object keys must be sorted.
    const keys = Object.keys(tree);
    keys.sort();
    const buffer: string[] = [];
    buffer.push("{");
    let notFirst = false;
    for (const key of keys) {
      if (notFirst) {
        buffer.push(",");
      }
      notFirst = true;
      buffer.push(JSON.stringify(key));
      buffer.push(":'");
      const valueHash = await merkle(algorithm, tree[key]);
      buffer.push(valueHash.base85);
      buffer.push("'");
    }
    buffer.push("}");
    canon = buffer.join("");
  }
  const utf8 = utf8Encoder.encode(canon);
  return await MerkleHash.deriveFrom(algorithm, utf8);
}

export { type DigestAlgorithmType };
